Fixes #27788 - IPv6 support for trusted hosts

lzap opened the pull request:
User can set IPv6 address in hostname format correctly, e.g.
`[::ffff:127.0.0.1]` however the matching code will not recognize this
syntax and will not match a proxy connecting through IPv6.

Added also a test cases for the cases.

lzap committed:
Fixes #27788 - IPv6 support for trusted hosts

timogoebel commented:
Looks good, I'm just wondering if we want to support network ranges as well. So a user can allow `192.168.1.0/24` or `[2001:beef::/48]` or something.

ekohl commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -122,7 +122,7 @@ def detect_matching_host(allowed_hosts, request_hosts)
         return allowed if allowed =~ rex
       else
-        return allowed if allowed == request
+        return allowed if (allowed == request) || (allowed == "[#{request}]")
```
Should we use `IPAddr` to try to parse it perhaps?
```
[16] pry(main)> IPAddr.new('[::ffff:127.0.0.1]').native().to_s
=> "127.0.0.1"
```

lzap replied:
Do you mean comparing IPAddr objects instead? Looks cleaner, however it can also lead to more complex code because input is mixture of hostnames and IPs.

ekohl replied:
Yes. It should probably be a fallback with a try/except. Since IPAddr does support CIDR notation, that could be useful as well to support rather than regexes.

lzap commented:
I don't mind supporting ranges, I need to take a closer look because this actually does not work with brackets and it required a different approach:

```
>> IPAddr.new('2001:beef::/48')
=> #<IPAddr: IPv6:2001:beef:0000:0000:0000:0000:0000:0000/ffff:ffff:ffff:0000:0000:0000:0000:0000>
>> IPAddr.new('[2001:beef::/48]')
IPAddr::InvalidAddressError (invalid address: [2001:beef::/48])
```

ekohl commented:
That's correct. CIDR with brackets is invalid anyway. Brackets are only useful when you need to add a port to it and CIDR ranges never have ports. That means we should never expect it as an input, but the setting can be a CIDR range.

timogoebel commented:
> Brackets are only useful when you need to add a port to it and CIDR ranges never have ports.

True. Why do we want brackets in the first place?

lzap commented:
> True. Why do we want brackets in the first place?

It's hostname array, IPv6 can go into hostname only with brackets. Anyway, I am currently working on allowing hostnames AND IP addresses in that array.

lzap commented:
Rebased, I had to rename the validator since it now matches hostnames AND addresses but it does not appear to be used anywhere else than for this setting so probably harmless. Added more tests for all possible cases.

ekohl commented on test/controllers/smart_proxy_auth_test.rb:
```diff
@@ -140,16 +140,56 @@ def test_certificate_with_sans_permits_access
+
+  def test_trusted_puppet_master_hosts_by_ipv6_subnet_match
+    Resolv.any_instance.expects(:getnames).with('2001:beef::1').returns([])
+    @request.env['REMOTE_ADDR'] = '2001:beef::1'
```
The `2001:db8::/32` range is reserved for documentation (tools.ietf.org/html/rfc3849` while `2001:beef::/32` is part of `2001:b000::/20` and assigned to an ISP in Australia.

```suggestion
    @request.env['REMOTE_ADDR'] = '2001:db8:beef::1'
```

lzap replied:
Thrilled to read whos 2001:cafe ;-)

ekohl commented on test/controllers/smart_proxy_auth_test.rb:
```diff
@@ -140,16 +140,56 @@ def test_certificate_with_sans_permits_access
+  def test_trusted_puppet_master_hosts_by_ipv6_subnet_match
+    Resolv.any_instance.expects(:getnames).with('2001:beef::1').returns([])
+    @request.env['REMOTE_ADDR'] = '2001:beef::1'
+    Setting[:trusted_hosts] = ['2001:beef::1/48']
```
The last `1` is not needed since there's a mask anyway:
```suggestion
    Setting[:trusted_hosts] = ['2001:db8:beef::/48']
```

lzap replied:
Yeah copy and paste error.

ekohl commented on app/validators/array_hostnames_ips_validator.rb:
```diff
@@ -0,0 +1,8 @@
+class ArrayHostnamesIpsValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    record.errors.add attribute, _("must contain valid hostnames") unless value.all? do |item|
+      item.to_s.match(URI::HOST) || IPAddr.new(item) rescue nil
```
If there's a `rescue nil`, doesn't that mean the check is always going to pass?

I'm also wondering about `item.to_s` on the URI matching but not on the `IPAddr.new`. I don't know what the data type is and it may just be my lack of Rails knowledge. Is this an issue?

lzap replied:
Good catch that is probably incorrect.

The `to_s` was required because `match` is only defined for String so that the code does not throw error on different classes. To be on the safe side, I will add that too.

lzap commented:
Nice catch, updated example IPv6 addresses.

timogoebel commented on app/validators/array_hostnames_validator.rb:
```diff
@@ -1,5 +0,0 @@
-class ArrayHostnamesValidator < ActiveModel::EachValidator
```
How about we keep this + the test so plugins can use this?

lzap replied:
Hai! Rebased.

lzap commented:
Added those two guys back.

tbrisker reviewed (dismissed):
Thanks @lzap !

tbrisker commented:
pushed rebase so tests run again

tbrisker reviewed (approved):
Thanks @lzap !
